### PR TITLE
DeepONet should support arbitrary trunk/branch networks

### LINKED ISSUE #71: DeepONet should support arbitrary trunk/branch networks
Currently, we implement the DeepONet with fully connected networks for branch and trunk. Depending on the problem, one would like to replace these networks by more appropriate ones (e.g. CNNs as branch network).

Generalize DeepONet to take branch/trunk network in constructor. Per default, it should use DeepResidualNetwork.

### PR DESCRIPTION
**Pull request by samuelburbulla**
# Feature: DeepONet should support arbitrary trunk/branch networks

### Which issue does this PR tackle?

  - Fixes #71 
 
### How does it solve the problem?

  - Adds possibility to use custom branch and trunk network in DeepONet by introducing `branch_network` and `trunk_network` arguments.

### How are the changes tested?

  - Added `test_custom_deeponet`.

## Checklist for Contributors

- [x] Scope: This PR tackles exactly one problem.
- [x] Conventions: The branch follows the `feature/title-slug` convention.
- [x] Conventions: The PR title follows the `Bugfix: Title` convention.
- [x] Coding style: The code passes all pre-commit hooks.
- [x] Documentation: All changes are well-documented.
- [x] Tests: New features are tested and all tests pass successfully.
- [x] Changelog: Updated CHANGELOG.md for new features or breaking changes.
- [x] Review: A suitable reviewer has been assigned.

## Checklist for Reviewers:

- [x] The PR solves the issue it claims to solve and only this one.
- [ ] Changes are tested sufficiently and all tests pass.
- [ ] Documentation is complete and well-written.
- [x] Changelog has been updated, if necessary.

### CODE REVIEW

File: src/continuiti/operators/deeponet.py [hunk 1]
```diff
@@ -30,8 +30,14 @@ class DeepONet(Operator):
...
         basis_functions: Number of basis functions.
-        act: Activation function.
+        act: Activation function used in default trunk and branch networks.
         device: Device.
+        branch_network: Custom branch network that maps input function
+            evaluations to `basis_functions` many coefficients (if set,
+            branch_width and branch_depth will be ignored).
+        trunk_network: Custom trunk network that maps `shapes.y.dim`-dimensional
+            evaluation coordinates to `basis_functions` many basis function
+            evaluations (if set, trunk_width and trunk_depth will be ignored).
```
Comment (JakobEliasWagner) [inline src/continuiti/operators/deeponet.py:40]: Not only `<net>_width`, and `<net>_depth` will be ignored but also the number of basis functions is dependent on the branch and trunk network. The output dimension of the of the trunk and branch networks is dim(v) and the number of basis functions.
`basis_functions` should be ignored as well and inferred from these networks if possible.

Comment (JakobEliasWagner) [reply src/continuiti/operators/deeponet.py:40]: Or is this a fallback for when only the branch-network is set? If this is the case an assert statement could underline this (checking if the number of basis functions in both branch and trunk network match).

File: tests/operators/test_deeponet.py
Comment (JakobEliasWagner) [inline tests/operators/test_deeponet.py:1]: An additional test covering custom trunk networks would be nice
```suggestion


@pytest.mark.slow
def test_custom_branch_network():
    # Data set
    n_sensors = 32
    dataset = SineBenchmark(n_train=1, n_sensors=n_sensors).train_dataset

    # CNN as branch network
    basis_functions = 8
    branch_network = torch.nn.Sequential(
        torch.nn.Conv1d(1, 16, kernel_size=3, padding=1),
        torch.nn.ReLU(),
        torch.nn.Conv1d(16, 1, kernel_size=3, padding=1),
        torch.nn.Flatten(),
        torch.nn.Linear(n_sensors, basis_functions),
    )

    # Operator
    operator = DeepONet(
        dataset.shapes,
        branch_network=branch_network,
        basis_functions=basis_functions,
    )

    # Train
    Trainer(operator).fit(dataset, tol=1e-3)

    # Check solution
    x, u, y, v = dataset.x, dataset.u, dataset.y, dataset.v
    assert MSELoss()(operator, x, u, y, v) < 1e-3


@pytest.mark.slow
def test_custom_trunk_network():
    # Data set
    n_sensors = 32
    dataset = SineBenchmark(n_train=1, n_sensors=n_sensors).train_dataset
    
    class SineAct(torch.nn.Module):
        def __init__(self):
            super().__init__()
            
        def forward(self, src: torch.tensor):
            return torch.sin(src)
    
    # CNN as trunk network
    basis_functions = 32
    trunk_network = torch.nn.Sequential(
        torch.nn.Linear(1, 32),
        torch.nn.LayerNorm(32),
        torch.nn.Tanh(),
        torch.nn.Linear(32, 32),
        torch.nn.LayerNorm(32),
        torch.nn.Tanh(),
        torch.nn.Linear(32, 32),
        torch.nn.LayerNorm(32),
        torch.nn.Tanh(),
        torch.nn.Linear(32, 32),
        torch.nn.LayerNorm(32),
        torch.nn.Tanh(),
        torch.nn.Linear(32, basis_functions),
    )

    # Operator
    operator = DeepONet(
        dataset.shapes,
        trunk_network=trunk_network,
        basis_functions=basis_functions,
    )

    # Train
    Trainer(operator).fit(dataset, tol=1e-3)

    # Check solution
    x, u, y, v = dataset.x, dataset.u, dataset.y, dataset.v
    assert MSELoss()(operator, x, u, y, v) < 1e-3

``

**Review by JakobEliasWagner — CHANGES_REQUESTED** [review]

**Review by JakobEliasWagner — APPROVED** [review]
lgtm
